### PR TITLE
Pro-keys Parsing

### PR DESCRIPTION
**Pull request by EliteAsian123**
WIP

### CODE REVIEW

File: YARG.Core/Chart/Loaders/MoonSong/MoonSongLoader.ProKeys.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+            // Disjoint chords
+            foreach (var note in moonNote.chord)
+            {
+                if (note.length != moonNote.length)
+                {
+                    flags |= ProKeysNoteFlags.Disjoint;
+                    break;
+                }
+            }
```
Comment (TheNathannator) [inline YARG.Core/Chart/Loaders/MoonSong/MoonSongLoader.ProKeys.cs:61]: Do we need to mark extended sustains and disjoint chords in Pro Keys? Mechanically, everything is treated as an individual note, and you can hit notes in chords independently from each other.

File: YARG.Core/Chart/SongChart.cs [hunk 2]
```diff
@@ -341,13 +344,18 @@ static double VoxMin(IEnumerable<VocalsTrack> tracks)
             double totalStartTime = 0;
 
             // Tracks
+
             totalStartTime = Math.Min(TrackMin(FiveFretTracks), totalStartTime);
             totalStartTime = Math.Min(TrackMin(SixFretTracks), totalStartTime);
             totalStartTime = Math.Min(TrackMin(DrumsTracks), totalStartTime);
             totalStartTime = Math.Min(TrackMin(ProGuitarTracks), totalStartTime);
+
+            totalStartTime = Math.Min(ProKeys.GetStartTime(), totalStartTime);
```
Comment (TheNathannator) [inline YARG.Core/Chart/SongChart.cs:353]:
```suggestion
            totalStartTime = Math.Min(TrackMin(ProKeys), totalStartTime);
```

File: YARG.Core/Chart/SongChart.cs [hunk 3]
```diff
@@ -369,13 +377,18 @@ static double VoxMax(IEnumerable<VocalsTrack> tracks)
             double totalEndTime = 0;
 
             // Tracks
+
             totalEndTime = Math.Max(TrackMax(FiveFretTracks), totalEndTime);
             totalEndTime = Math.Max(TrackMax(SixFretTracks), totalEndTime);
             totalEndTime = Math.Max(TrackMax(DrumsTracks), totalEndTime);
             totalEndTime = Math.Max(TrackMax(ProGuitarTracks), totalEndTime);
+
+            totalEndTime = Math.Max(ProKeys.GetEndTime(), totalEndTime);
```
Comment (TheNathannator) [inline YARG.Core/Chart/SongChart.cs:386]:
```suggestion
            totalEndTime = Math.Max(TrackMax(ProKeys), totalEndTime);
```

File: YARG.Core/Chart/SongChart.cs [hunk 4]
```diff
@@ -397,13 +410,18 @@ static uint VoxMin(IEnumerable<VocalsTrack> tracks)
             uint totalFirstTick = 0;
 
             // Tracks
+
             totalFirstTick = Math.Min(TrackMin(FiveFretTracks), totalFirstTick);
             totalFirstTick = Math.Min(TrackMin(SixFretTracks), totalFirstTick);
             totalFirstTick = Math.Min(TrackMin(DrumsTracks), totalFirstTick);
             totalFirstTick = Math.Min(TrackMin(ProGuitarTracks), totalFirstTick);
+
+            totalFirstTick = Math.Min(ProKeys.GetFirstTick(), totalFirstTick);
```
Comment (TheNathannator) [inline YARG.Core/Chart/SongChart.cs:419]:
```suggestion
            totalFirstTick = Math.Min(TrackMin(ProKeys), totalFirstTick);
```

File: YARG.Core/Chart/SongChart.cs [hunk 5]
```diff
@@ -425,13 +443,18 @@ static uint VoxMax(IEnumerable<VocalsTrack> tracks)
             uint totalLastTick = 0;
 
             // Tracks
+
             totalLastTick = Math.Max(TrackMax(FiveFretTracks), totalLastTick);
             totalLastTick = Math.Max(TrackMax(SixFretTracks), totalLastTick);
             totalLastTick = Math.Max(TrackMax(DrumsTracks), totalLastTick);
             totalLastTick = Math.Max(TrackMax(ProGuitarTracks), totalLastTick);
+
+            totalLastTick = Math.Max(ProKeys.GetLastTick(), totalLastTick);
```
Comment (TheNathannator) [inline YARG.Core/Chart/SongChart.cs:452]:
```suggestion
            totalLastTick = Math.Max(TrackMax(ProKeys), totalLastTick);
```

File: YARG.Core/MoonscraperChartParser/IO/Midi/MidReader.ProcessLists.cs [hunk 6]
```diff
@@ -80,6 +81,13 @@ private struct CommonPhraseSettings
             lanePhrases = false,
         };
 
+        private static readonly CommonPhraseSettings ProKeysPhraseSettings = new()
+        {
+            soloNote = MidIOHelper.SOLO_NOTE,
```
Comment (TheNathannator) [inline YARG.Core/MoonscraperChartParser/IO/Midi/MidReader.ProcessLists.cs:86]: Pro Keys uses note 115 for marking solos, not note 103.

File: YARG.Core/MoonscraperChartParser/IO/Midi/MidReader.ProcessLists.cs [hunk 7]
```diff
@@ -80,6 +81,13 @@ private struct CommonPhraseSettings
...
+            versusPhrases = true,
```
Comment (TheNathannator) [inline YARG.Core/MoonscraperChartParser/IO/Midi/MidReader.ProcessLists.cs:87]: This should technically be false, the versus modes were removed in RB3 and Pro Keys does not contain them.

**Review by TheNathannator — CHANGES_REQUESTED** [review]
Some nitpicks, but otherwise looks good!

File: YARG.Core/Chart/SongChart.cs (hunk 2 above)
Comment (EliteAsian123) [reply YARG.Core/Chart/SongChart.cs:353]: `TrackMin` takes in an enumerable of instrument tracks, so I just used its logic on the single pro-keys track.

File: YARG.Core/MoonscraperChartParser/IO/Midi/MidReader.ProcessLists.cs (hunk 7 above)
Comment (EliteAsian123) [reply YARG.Core/MoonscraperChartParser/IO/Midi/MidReader.ProcessLists.cs]: Fixed

File: YARG.Core/MoonscraperChartParser/IO/Midi/MidReader.ProcessLists.cs (hunk 6 above)
Comment (EliteAsian123) [reply YARG.Core/MoonscraperChartParser/IO/Midi/MidReader.ProcessLists.cs]: Fixed

File: YARG.Core/Chart/Loaders/MoonSong/MoonSongLoader.ProKeys.cs (hunk 1 above)
Comment (EliteAsian123) [reply YARG.Core/Chart/Loaders/MoonSong/MoonSongLoader.ProKeys.cs]: Good point. I just removed those flags entirely.

File: YARG.Core/Chart/SongChart.cs (hunk 2 above)
Comment (TheNathannator) [reply YARG.Core/Chart/SongChart.cs:353]: Ah right, seems I missed it was using the enumerable helpers and not the individual tracks lol

**Pull request merged**